docs/internal/contributing/how-to-convert-config-to-per-tenant-limit.md

```diff
@@ -0,0 +1,181 @@
+
+   ```go
+   ResultsCacheForUnalignedQueryEnabled bool `yaml:"cache_unaligned_requests" json:"cache_unaligned_requests" category:"advanced"`
+   ```
+
+2. Add the binding of the CLI flag to the Limits config.
+
+   Again, group it with other related options.
+   Ensure the same default value is maintained.
+
+   In `validation.Limits` struct's `RegisterFlags` method:
+
+   ```go
+   f.BoolVar(&l.ResultsCacheForUnalignedQueryEnabled, "query-frontend.cache-unaligned-requests", false, "Cache requests that are not step-aligned.")
+   ```
+
+3. Expose the tenant-specific overrides for the new config option
+
+   First add the tenant-specific method and docstring to the _interface_ which uses the config option.
+   The interface definition - which may need to be created - should be named `Limits` and define all related per-tenant config options.
+
+   In the example PR, the config option was used in `pkg/frontend/querymiddleware/limits.go`.
+   Interface `querymiddleware.Limits` already existed, so we only needed to add a new interface method:
+
+   ```go
+    // ResultsCacheForUnalignedQueryEnabled returns whether to cache results for queries that are not step-aligned
+    ResultsCacheForUnalignedQueryEnabled(userID string) bool
+   ```
+
+   This interface method may also need to be added to test mocks.
+
+   Then add the implementation of the interface method to the `validation.Overrides` struct in `pkg/util/validation/limits.go`:
+
+   ```go
+   func (o *Overrides) ResultsCacheForUnalignedQueryEnabled(userID string) bool {
+       return o.getOverridesForUser(user).ResultsCacheForUnalignedQueryEnabled
+   }
+   ```
+
+   This ensures that the `Overrides` struct, which holds any tenant-specific runtime config, will implement the `Limits` interface needed for your codepath.
+
+4. Decide how to consume the option when multi-tenant paths have different limits.
+
+   Multi-tenant codepaths using limits can involve a mix of tenants using limits and tenants which have varying tenant-specific overrides.
+   These codepaths may need to determine whether to take a largest, smallest, or other value from the varied tenant limits.
+
+   In the example PR, we decided to only cache the request if all tenants have the caching enabled.
+   In `pkg/util/validation/limits.go`, we added a helper function:
+
+   ```go
+    // AllTrueBooleansPerTenant returns true only if limit func is true for all given tenants
+    func AllTrueBooleansPerTenant(tenantIDs []string, f func(string) bool) bool {
+        for _, tenantID := range tenantIDs {
+            if !f(tenantID) {
+                return false
+            }
+        }
+        return true
+    }
+   ```
+
+   Similar helper functions may already exist for similar purposes: `SmallestPositiveNonZeroIntPerTenant`, `LargestPositiveNonZeroDurationPerTenant`, etc.
+
+5. Consume the overrides with the multiple tenant limits.
```

Comment:
Using the limits from multiple tenants at once is specific to the read path. Can you note that here or mark this step as optional?